### ISSUE TITLE
Document Reviewer Web- New batch download, stitch logic implementation

### ISSUE THREAD

**Issue by abin-aot** [bcgov/foi-flow#4680]
Title of ticket: 

#### Description
This task has been created from the observations of the #3694 (Performance Testing , optimization and bench marking on DOc. Reviewer Web)  and it was decided to release the second stage of performance optimization as part of 9.1 release, since  first set of benchmarking is acceptable for the business! ( @lmullane , @m-prodan ) - please review the `LAST` set of comments on #3694. 

The purpose of this task is to integrate the proven logic (POC attached) to Document Reviewer Web, which to batch downloads, stitch the docs, pages as per the sorted fashion(modified date), if downloaded.  Bench Marking of this logic will cover the CFD, MSD personal scanning scenarios, Also the large file sets above 1000 pages, 250 documents and sizes above 300 MB. Idea is to load , stitch the pages quickly compared with the existing logic. Also, will need to optimize the annotation loading, redlining issue with API, if any.

#### POC CODE 
https://bcgov.sharepoint.com/:u:/r/teams/03521/Shared%20Documents/Development/poc_batchdownload_%20pdftron.zip?csf=1&web=1&e=pAQFg5

#### NOT IN SCOPE

- Any new UX improvisation for better Document loading, Say suggestion like `DOCUMENT SETS` etc. as per #3694. This will be a new story and task, if required.

#### Dependencies
Are there any dependencies? 
999+ since page document not stitching , https://support.apryse.com/support/tickets/64255

#### DOD
- [x] Compare the performance of new logic with previous bench mark on TEST MARSHAL, [EDU-2023-18090937](https://test-marshal-foirequest.apps.silver.devops.gov.bc.ca/foi/foirequests/655/ministryrequest/655) took **1.44mins**.
 https://dev-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/584 - Similar set took just **45 s**.
- [x] Compare the performance of new logic with previous bench mark on TEST MARSHAL.
dev-marshal request [BBB-2023-27100403] with **214 files, 4662 pages with 600+ annotations** took around **4 mins**
https://dev-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/656
test-marshal request[BBB-2023-30101149] with same set of files took around **8 mins**
https://test-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/721
- [x] Compare, Analyze, Benchmark the loading time  of 500-700 records, 300 MB with 2000 pages - Check the room for optimization 
- [ ] Compare the loading time  of 800-1000 records, 400-500 MB with 4000 pages -  Check the room for optimization 
- [x]  Add the annotations on to the above load conditions and compare the loading time and operations in generals - Search comments, annotation, navigations, redlining, response packages
- [x]  Analyze , implement, bench mark  the POC logic to the SAVE REDLINE, RESPONSE package . IMportant : REDLINE performance need to be checked with Single REDLINE doc. as well.
- [ ]  Coordinate , inform the new bench mark to business, team
- [ ]  Did we checked the CFD, MSD scenarios with large sets(1000+) individual  one page pdf files, if PDFTRON fixed the bug mentioned . https://support.apryse.com/support/tickets/64255
- [x]  Merge test-marshal changes to dev-marshal
- [x]  Merge latest main changes to dev-marshal
- [x]  Compare the loading time  of 800-900 records with 2KB files(total 800 - 900 pages) each.
request [BBB-2023-01111026] - dev-marshal load time - less than 2 mins (800 files) - less than 3 mins (900 files)
https://dev-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/657
https://dev-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/658
test-marshal load time - more than 8 mins (800 files) - more than 12 mins (900 files)
https://test-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/731
https://test-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/732

**Comment by aparna-aot** [comment]
**Status & Findings :**
Status:
- Initial Stitching with slicer logic is finished.
- Redline download with slicer logic - Was unable to implement slicer logic for multiple divisions so implemented the logic for REDLINE SINGLE STITCH & SINGLE DIVISION REDLINE SCENARIO.
More Details of the issue faced : Issue was with async nature of calls. For 2 divisions, the last document of the first division was entering the useEffect of pdftronDocObjectsForRedline variable only after starting the next division loop messing up the rest of stitching.
So now I added a condition : if(res.issingleredlinepackage == 'Y' || divisions.length == 1) -> Here new slicer method is called (stitchSingleDivisionRedlineExport) else -> previous method without slicer (stitchForRedlineExport).

Finding:

- Initial stitching was working for all requests with improved performance noted above with 10.4 version of pdftron. However after merging the MAIN branch changes to dev-marshal, I'm getting errors for some of the requests in dev-marshal(not always), however I made Divya and Nimya try loading those documents and it was working for them. So some kind of a sporadic issue. So might need to look into it. Attaching the screenshot of the error with request details:
Request urls - https://dev-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/656https://dev-reviewer-web-marshal.apps.silver.devops.gov.bc.ca/foi/660[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBL2dHQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--433a46cd73cfdbdf5f2069cd4a8d5372f594d130/Docreviewer-error.png)

**Comment by nimya-aot** [comment]
Working good with 170 documents on dev-marshal

**Comment by abin-aot** [comment]
Client side stitching and Redline stitching is dependent on the new slicer logic. Since the outstanding issue, https://support.apryse.com/support/tickets/64255 is not ready at the time we were ready to ship to DEV, TEST, PROD, we need to analyze their new logic and integrate with Document reviewer. Means, there is a limit of # of individual documents(not pages!) that can stitched - which 170 to 999+ depends on the size of the document. Loading the document dont have lower limit issue of 170. Lower limit issue is with Redline download.
